File: leetcode/golang/src/delete-and-earn.go
/* -------------------------------------------------------------------------- */
/*                https://leetcode.com/problems/delete-and-earn               */
/* -------------------------------------------------------------------------- */
package main

func DeleteAndEarn(nums []int) int {
	max := nums[0]
	for i := 1; i < len(nums); i++ {
		if nums[i] > max {
			max = nums[i]
		}
	}

	freq := make([]int, max+1)
	for i := 0; i < len(nums); i++ {
		freq[nums[i]] += 1
	}

	for i := 2; i < max+1; i++ {
		freq[i] = i*freq[i] + freq[i-2]
		if freq[i-1] > freq[i] {
			freq[i] = freq[i-1]
		}
	}

	return freq[max]
}

File: leetcode/rust/src/delete_and_earn.rs
/* -------------------------------------------------------------------------- */
/*                https://leetcode.com/problems/delete-and-earn               */
/* -------------------------------------------------------------------------- */

pub fn delete_and_earn(nums: Vec<i32>) -> i32 {
    let max = *nums.iter().max().unwrap() as usize;

    let mut freq = vec![0i32; max + 1];
    nums.iter().for_each(|x| freq[*x as usize] += 1);

    for i in 2..(max + 1usize) {
        freq[i] = std::cmp::max(i as i32 * freq[i] + freq[i - 2], freq[i - 1]);
    }

    return *freq.last().unwrap();
}
